File: numerical_analysis/newton.py
import numpy as np

class Newton():

    def func(self , x):
        return x**2 - np.sin(x)
    
    def deriv_func(self,x):
        return 2*x - np.cos(x)
    
    def newton(self , x , erro):

        #h = self.func(x)/self.deriv_func(x)
        iter = 0

        while True:
            h = self.func(x)/self.deriv_func(x)

            print("Iteração número", iter , "Com a iteração em ", h)

            if np.abs(h) < erro:
                break

            x = x-h
            iter +=1
        
        print("Valor da raiz é ", self.func(h))


def main():
    print("Calculando a raiz de X*X - sin(x) ")

    x = np.float128(input("Digite o X: "))
    erro = np.float128(input("Digite o erro: "))

    new = Newton()

    new.newton(x,erro)

if __name__ == "__main__":
    main()

File: numerical_analysis/newton.cpp
#include <iostream>
#include<cmath>
 
#include<boost/multiprecision/cpp_dec_float.hpp>

namespace mp = boost::multiprecision;
            
double func(double x){
  return x*x - sin(x);
}

double derivada(double x){
  return 2*x - cos(x);
}

void newton(double x, mp::cpp_dec_float_50 erro ){

  double h = func(x) / derivada(x);
  int iter =0;
  while(fabs(h) > erro){
    h = func(x)/derivada(x);

    std::cout<<"Iteração número: "<<iter <<"Com o erro: "<< h << std::endl;

    //fazendo x+1 
    x = x - h;
    iter++;
  }

  std::cout<< "Valor da raiz é :" << x;

}

int main(){
  
  std::cout<<"Digite o erro: ";
  mp::cpp_dec_float_50 erro;
  std::cin>>erro;

  //std::cout<<"\n Digite os intervalos: ";
  //double a ,b;
  //std::cin>>a >>b;
  std::cout<<"\n\n Digite o X: ";
  double x;
  std::cin>>x;

  newton(x , erro);
  return 0;
}
